In [1]:

import os, types
import pandas as pd
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# @hidden_cell
# The following code accesses a file in your IBM Cloud Object Storage. It includes your credentials.
# You might want to remove those credentials before you share the notebook.

if os.environ.get('RUNTIME_ENV_LOCATION_TYPE') == 'external':
    endpoint_1f4d4c4de89540c5b91283c1fbf2e8e1 = 'https://s3.us.cloud-object-storage.appdomain.cloud'
else:
    endpoint_1f4d4c4de89540c5b91283c1fbf2e8e1 = 'https://s3.private.us.cloud-object-storage.appdomain.cloud'

client_1f4d4c4de89540c5b91283c1fbf2e8e1 = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='dDuT9Y7UtbO9-WmawNyBuALgL4gYUI6T2MuRzIvMcRxJ',
    ibm_auth_endpoint="https://iam.cloud.ibm.com/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url=endpoint_1f4d4c4de89540c5b91283c1fbf2e8e1)

body = client_1f4d4c4de89540c5b91283c1fbf2e8e1.get_object(Bucket='latihanwatsonstudio-donotdelete-pr-o35aesrtn21y0t',Key='Iris.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_1 = pd.read_csv(body)
df_data_1.head()


Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

# Data Preparation

In [3]:
# Normalization
'''
Normalization adalah salah satu teknik yang dipakai dalam data preparation. Tujuan dari 
normalisasi adalah mengubah nilai-nilai dari sebuah fitur ke dalam skala yang sama. 
Normalization memungkinkan kenaikan performa dan stabilitas dari sebuah model machine learning.
(Gaji dan Umur, terkecil = 0 dan terbesar = 1)

Info : https://scikit-learn.org/0.16/modules/generated/sklearn.preprocessing.MinMaxScaler.html
       https://www.dicoding.com/academies/184/tutorials/8352?from=8347
'''

from sklearn.preprocessing import MinMaxScaler
data = [[12000000, 33], [35000000, 45], [4000000, 23], [6500000, 26], [9000000, 29]]
scaler = MinMaxScaler()
scaler.fit(data)
print(scaler.transform(data))

[[0.25806452 0.45454545]
 [1.         1.        ]
 [0.         0.        ]
 [0.08064516 0.13636364]
 [0.16129032 0.27272727]]


In [7]:
# Standardization
'''
Standardization adalah proses konversi nilai-nilai dari suatu fitur sehingga nilai-nilai 
tersebut memiliki skala yang sama. Z score adalah metode paling populer untuk standardisasi 
di mana setiap nilai pada sebuah atribut numerik akan dikurangi dengan rata-rata dan dibagi 
dengan standar deviasi dari seluruh nilai pada sebuah kolom atribut.

Info : https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html
       https://www.dicoding.com/academies/184/tutorials/8352?from=8347
'''

from sklearn import preprocessing
data = [[12000000, 33], [35000000, 45], [4000000, 23], [6500000, 26], [9000000, 29]]
scaler = preprocessing.StandardScaler().fit(data)
data = scaler.transform(data)
data

array([[-0.11638732,  0.23521877],
       [ 1.94277296,  1.80334389],
       [-0.83261698, -1.07155217],
       [-0.60879521, -0.67952089],
       [-0.38497344, -0.28748961]])

# Datasets

Pemisahan data training dan data test

- Data test = 20%
- Data training = 80%

- train_test_split = mengacak (shuffling) datasets agar akurasi ML optimal
- random_state (umumnya 0, 1 atau 42) = untuk random seed proses shuffling (pengacakan)

In [5]:
# Perbedaan penetuan random_state
from sklearn.model_selection import train_test_split
 
X_data = range(10)
y_data = range(10)
 
print("random_state ditentukan (benar)")
for i in range(3):
    X_train, X_test, y_train, y_test = train_test_split(X_data, y_data, test_size = 0.3, random_state = 42)
    print(y_test)
 
 
print("random_state tidak ditentukan (salah)")
for i in range(3):
    X_train, X_test, y_train, y_test = train_test_split(X_data, y_data, test_size = 0.3, random_state = None)
    print(y_test)

random_state ditentukan (benar)
[8, 1, 5]
[8, 1, 5]
[8, 1, 5]
random_state tidak ditentukan (salah)
[0, 8, 1]
[8, 7, 6]
[0, 7, 1]


# Latihan SKLearn Train Test Split

- Diperlukan pembagian data untuk train set (+- 80%) dan test set (+- 20%)
- Analogi buah (Atribut dan Label)
    - Atribut : warnanya kuning, bentuknya memanjang, ukurannya segenggaman tangan
    - Label : Pisang

In [6]:
import sklearn
from sklearn import datasets
from sklearn.model_selection import train_test_split

'''
x_test = Atribut Test set
y_test = Target (Label) Test Set
x_train = Atribut Train Set
y_train = Target (Label) Train Set
'''


# load iris dataset (jumlah data = 150)
iris = datasets.load_iris()

# pisahkan atribut dan label pada iris dataset
x=iris.data
y=iris.target

# membagi dataset menjadi training dan testing (Data Test = 20%)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

# menghitung panjang/jumlah data pada x_test (output = 30)
len(x_test)

30

# Latihan SKLearn Cross Validation Split

- Dari 80% train set, beberapa persen datanya perlu dijadikan validation set.
  Validation set atau holdout validation adalah bagian dari train set yang 
  dipakai untuk pengujian model pada tahap awal
- Salah satu cara untuk mendapatkan validation set adalah dengan K-Fold Cross Validation atau lebih sering disebut cross validation

In [3]:
import sklearn
from sklearn import datasets

# Load iris dataset
iris = datasets.load_iris()

In [4]:
# mendefinisikan atribut dan label pada dataset
x=iris.data
y=iris.target

In [5]:
from sklearn import tree
 
# membuat model dengan decision tree ckassifier
clf = tree.DecisionTreeClassifier()

In [7]:
from sklearn.model_selection import cross_val_score
 
# mengevaluasi performa model dengan cross_val_score
# clf = Model machine laerning
# x = atribut
# y = label
# cv = jumlah flod (lipatan) yang akan digunakan untuk cross validation
scores = cross_val_score(clf, x, y, cv=5)
scores

array([0.96666667, 0.96666667, 0.9       , 0.96666667, 1.        ])

Jika yang dijadikan validation set adalah :
- Fold pertama, akurasi = 0.96666667
- Flod kedua, akurasi = 0.96666667
- Fold ketiga, akurasi = 0,9
- dst.

Jika akurasinya bervariasi di kisaran 0.85 sampai 0.99 maka
model tersebut dapat dikatakan baik